lib/open_project/scm/adapters/git.rb

```diff
@@ -0,0 +1,368 @@
+            )
+          end
+        end
+
+        def encode_full_path(name, path)
+          p = scm_encode(@path_encoding, 'UTF-8', path || '')
+          full_path = p.empty? ? name : "#{p}/#{name}"
+          scm_encode('UTF-8', @path_encoding, full_path)
+        end
+
+        def lastrev(path, rev)
+          return nil if path.nil?
+          args = %w|log --no-color --encoding=UTF-8 --date=iso --pretty=fuller --no-merges -n 1|
+          args << rev if rev
+          args << '--' << path unless path.empty?
+          lines = capture_git(args).lines
+          begin
+            build_lastrev(lines)
+          rescue NoMethodError
+            logger.error("The revision '#{path}' has a wrong format")
+            return nil
+          end
+        end
+
+        def build_lastrev(lines)
+          id = lines[0].split[1]
+          author = lines[1].match('Author:\s+(.*)$')[1]
+          time = Time.parse(lines[4].match('CommitDate:\s+(.*)$')[1])
+
+          Revision.new(
+            identifier: id,
+            scmid: id,
+            author: author,
+            time: time,
+            message: nil,
+            paths: nil
+          )
+        end
+
+        def revisions(path, identifier_from, identifier_to, options = {})
+          revisions = Revisions.new
+          args = build_revision_args(path, identifier_from, identifier_to, options)
+
+          files = []
+          changeset = {}
+          parsing_descr = 0 # 0: not parsing desc or files, 1: parsing desc, 2: parsing files
+          parse_by_line(args, binmode: true) do |line|
+            if line =~ /^commit ([0-9a-f]{40})$/
+              key = 'commit'
+              value = $1
+              if parsing_descr == 1 || parsing_descr == 2
+                parsing_descr = 0
+                revision = Revision.new(
+                  identifier: changeset[:commit],
+                  scmid: changeset[:commit],
+                  author: changeset[:author],
+                  time: Time.parse(changeset[:date]),
+                  message: changeset[:description],
+                  paths: files
+                )
+                if block_given?
+                  yield revision
+                else
+                  revisions << revision
+                end
+                changeset = {}
+                files = []
+              end
+              changeset[:commit] = $1
+            elsif (parsing_descr == 0) && line =~ /^(\w+):\s*(.*)$/
+              key = $1
+              value = $2
+              if key == 'Author'
+                changeset[:author] = value
+              elsif key == 'CommitDate'
+                changeset[:date] = value
+              end
+            elsif (parsing_descr == 0) && line.chomp.to_s == ''
+              parsing_descr = 1
+              changeset[:description] = ''
+            elsif (parsing_descr == 1 || parsing_descr == 2) &&
+              (line =~ /^:\d+\s+\d+\s+[0-9a-f.]+\s+[0-9a-f.]+\s+(\w)\t(.+)$/)
```

Comment:
Align the operands of a condition in a `elsif` statement spanning multiple lines.